cocos/platform/android/jni/JniHelper.h

```diff
@@ -72,159 +69,75 @@ class CC_DLL JniHelper
-                                       const std::string& methodName, 
-                                       Ts... xs) {
-        Vec3 ret;
-        cocos2d::JniMethodInfo t;
-        std::string signature = "(" + std::string(getJNISignature(xs...)) + ")[F";
-        if (cocos2d::JniHelper::getStaticMethodInfo(t, className.c_str(), methodName.c_str(), signature.c_str())) {
-            LocalRefMapType localRefs;
-            jfloatArray array = (jfloatArray) t.env->CallStaticObjectMethod(t.classID, t.methodID, convert(localRefs, t, xs)...);
-            jsize len = t.env->GetArrayLength(array);
-            if (len == 3) {
-                jfloat* elems = t.env->GetFloatArrayElements(array, 0);
-                ret.x = elems[0];
-                ret.y = elems[1];
-                ret.z = elems[2];
-                t.env->ReleaseFloatArrayElements(array, elems, 0);
-            }
-            t.env->DeleteLocalRef(t.classID);
-            deleteLocalRefs(t.env, localRefs);
-        } else {
-            reportError(className, methodName, signature);
-        }
-        return ret;
+    static inline double callStaticDoubleMethod(const std::string& className, 
+                                                const std::string& methodName, 
+                                                const Ts&... xs) {
+        return callStaticMethod<double>(className.c_str(), methodName.c_str(), xs...); 
     }
 
     template <typename... Ts>
-    static double callStaticDoubleMethod(const std::string& className, 
-                                         const std::string& methodName, 
-                                         Ts... xs) {
-        jdouble ret = 0.0;
-        cocos2d::JniMethodInfo t;
-        std::string signature = "(" + std::string(getJNISignature(xs...)) + ")D";
-        if (cocos2d::JniHelper::getStaticMethodInfo(t, className.c_str(), methodName.c_str(), signature.c_str())) {
-            LocalRefMapType localRefs;
-            ret = t.env->CallStaticDoubleMethod(t.classID, t.methodID, convert(localRefs, t, xs)...);
-            t.env->DeleteLocalRef(t.classID);
-            deleteLocalRefs(t.env, localRefs);
+    static inline std::string callStaticStringMethod(const std::string& className, 
+                                                     const std::string& methodName, 
+                                                     const Ts&... xs) {
+        return callStaticMethod<std::string>(className.c_str(), methodName.c_str(), xs...); 
+    }
+
+    template <typename Ret, typename... Args>
+    static Ret callStaticMethod(const char* className, const char* methodName, const Args& ...args) {
+        typedef typename JniHelperDetail::SignatureParser<Ret (Args...)>::Result SignatureSequence;
+        const char* signature = SignatureSequence::value;
+        JniMethodInfo t;
+        if (getStaticMethodInfo(t, className, methodName, signature)) {
+            LocalRefWrapper clazz(t.env, t.classID);
```

Comment:
clazz is not used, why need it?

Reply:
Got it, it is used to delete local ref, but it is a strange usage.

Reply:
LocalRefWrapper is used to delete local ref.
If we do not use it, we should declare a variable to hold the return value, then delete `t.classID`, and then returns.
But for case **return type==void** it is not allowed to declare a void variable.
So I use `LocalRefWrapper`'s **destructor** to delete `t.classID`

Reply:
May be should rename it similar to `lock_guard` to make it more clear: http://en.cppreference.com/w/cpp/thread/lock_guard